sdk/src/main/java/io/singularitynet/sdk/client/Sdk.java

```diff
@@ -13,71 +14,157 @@
+    /**
+     * New SDK instance with all dependencies provided manually.
+     * @param web3j web3j instance.
+     * @param ipfs IPFS instance.
+     * @param identity identity instance.
+     * @param registry Registry contract instance.
+     * @param mpe MultiPartyEscrow contract instance.
+     */
+    public Sdk(Web3j web3j, IPFS ipfs, Identity identity, Registry registry,
             MultiPartyEscrow mpe) {
         this.web3j = web3j;
         this.ipfs = ipfs;
-        this.signer = signer;
+        this.identity = identity;
         this.registry = registry;
         this.mpe = mpe;
+
+        this.ethereum = new Ethereum(web3j);
+        this.mpeContract = new MultiPartyEscrowContract(web3j, mpe);
+        this.metadataStorage = new IpfsMetadataStorage(ipfs);
+        this.registryContract = new RegistryContract(registry);
+        this.paymentChannelManager = new MpePaymentChannelManager(mpeContract);
     }
 
+    // TODO: move BaseServiceClient related initialization into
+    // BaseServiceClient constructor.
+    /**
+     * Return new instance of the ServiceClient for the given service.
+     * @param orgId organization id.
+     * @param serviceId service id.
+     * @param endpointGroupName name of the enpoint group to connect.
+     * @param paymentStrategy payment strategy to use.
+     * @return new instance of service client.
+     */
     public ServiceClient newServiceClient(String orgId, String serviceId,
             String endpointGroupName, PaymentStrategy paymentStrategy) {
         log.info("Start service client, orgId: {}, serviceId: {}, endpointGroupName: {}, paymentStrategy: {}",
                 orgId, serviceId, endpointGroupName, paymentStrategy);
 
-        RegistryContract registryContract = new RegistryContract(registry);
-        MultiPartyEscrowContract mpeContract = new MultiPartyEscrowContract(mpe);
+        MetadataProvider metadataProvider = getMetadataProvider(orgId, serviceId);
 
-        MetadataStorage metadataStorage = new IpfsMetadataStorage(ipfs);
-        MetadataProvider metadataProvider = new RegistryMetadataProvider(
-                orgId, serviceId, registryContract, metadataStorage);
-
-        DaemonConnection connection = new FirstEndpointDaemonConnection(
+        DaemonConnection connection = new RandomEndpointDaemonConnection(
                 endpointGroupName, metadataProvider);
+
         PaymentChannelStateService stateService = new PaymentChannelStateService(
-                connection, mpeContract, web3j, signer);
-        PaymentChannelProvider paymentChannelProvider = new
-            AskDaemonFirstPaymentChannelProvider(web3j, mpeContract,
-                    stateService);
+                connection, mpeContract.getContractAddress(), ethereum, identity);
+        PaymentChannelStateProvider paymentChannelStateProvider =
+            new AskDaemonFirstPaymentChannelProvider(mpeContract, stateService);
```

Comment:
should AskDaemonFirstPaymentChannelProvider also be part of daemon package as PaymentChannelStateService?
Also, will it make sense to hide state service, connection details from client?AskDaemonFirstPaymentChannelProvider can take care of all those details.

Reply:
> should AskDaemonFirstPaymentChannelProvider also be part of daemon package as PaymentChannelStateService?

Yes, I thought about it as well. After moving `Payment` into `payment` package it can be moved, I will add this.

> Also, will it make sense to hide state service, connection details from client?AskDaemonFirstPaymentChannelProvider can take care of all those details.

Not sure what do you mean. Could you please explain further?